### PR TITLE
TINY-9053: Exposing more types of TextPatterns 

### PR DESCRIPTION
**Pull request by ephox-mogran**
Related Ticket: TINY-9053

Description of Changes:
* exposed more types through TextPatterns in PublicApi
* created a test that uses just PublicApi to check that managing text pattern lookups are more managable

Questions
- does this need documentation changes? 
- is there a precedent for how verbose our changelog entry is for this? Does it need to list every type exposed? 

Pre-checks:
* [x] Changelog entry added
* [X] Tests have been added (if applicable)
* [X] Branch prefixed with `feature/`, `hotfix/` or `spike/`

Review:
* [X] Milestone set
* [x] Docs ticket created (if applicable)

GitHub issues (if applicable):

### CODE REVIEW

File: modules/tinymce/src/core/main/ts/api/textpatterns/TextPatterns.ts [hunk 1]
```diff
@@ -1,7 +1,14 @@
-import { RawPattern, RawDynamicPatternsLookup, DynamicPatternContext } from '../../textpatterns/core/PatternTypes';
+import { RawPattern, RawDynamicPatternsLookup, DynamicPatternContext, DynamicPatternsLookup, Pattern, BlockCmdPattern, BlockFormatPattern, InlineCmdPattern, InlinePattern, InlineFormatPattern } from '../../textpatterns/core/PatternTypes';
 
 export {
+  Pattern,
```
Comment (ephox-mogran) [inline modules/tinymce/src/core/main/ts/api/textpatterns/TextPatterns.ts:4]: I'm not sure if this is too many. Do we have a general reason to ever suppress types that are part of public API functions?

**Comment by lnewson** [comment]
> does this need documentation changes?

Types in general probably need documentation. For now I'd say no, but see what others say.

> is there a precedent for how verbose our changelog entry is for this? Does it need to list every type exposed?

In this case the bug was never released as it was only added in 6.2 so there should be no changelog entry for this.

File: modules/tinymce/src/core/main/ts/api/textpatterns/TextPatterns.ts [hunk 2]
```diff
@@ -1,7 +1,14 @@
...
   RawPattern,
+  DynamicPatternsLookup,
   RawDynamicPatternsLookup,
-  DynamicPatternContext
+  DynamicPatternContext,
```
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/api/textpatterns/TextPatterns.ts:8]: The incorrect type on `block` hasn't been fixed in this type by the looks of it. Is that expected?

Comment (ephox-mogran) [reply modules/tinymce/src/core/main/ts/api/textpatterns/TextPatterns.ts:8]: Nope. Completely forgot that part. Got carried away with exposing types. Please hold.

Comment (ephox-mogran) [reply modules/tinymce/src/core/main/ts/api/textpatterns/TextPatterns.ts:8]: That should be added now.

**Review by lnewson — APPROVED** [review]

File: modules/tinymce/src/core/main/ts/api/textpatterns/TextPatterns.ts [hunk 3]
```diff
@@ -1,7 +1,14 @@
...
+  BlockCmdPattern,
```
Comment (lnewson) [inline modules/tinymce/src/core/main/ts/api/textpatterns/TextPatterns.ts:9]: Should `BlockPattern` also be exposed here given we're exposing the `InlinePattern` type which is equivalent.

Comment (ephox-mogran) [reply modules/tinymce/src/core/main/ts/api/textpatterns/TextPatterns.ts:9]: Yeah. I think so for consistency. I'll update.

Comment (ephox-mogran) [reply modules/tinymce/src/core/main/ts/api/textpatterns/TextPatterns.ts:9]: Exposed BlockPattern and tested it was exposed.

File: modules/tinymce/src/core/main/ts/textpatterns/core/PatternTypes.ts [hunk 4]
```diff
@@ -53,7 +53,7 @@ export type Pattern = InlinePattern | BlockPattern;
 
 export interface DynamicPatternContext {
   readonly text: string; // the string from the start of the block to the cursor
-  readonly block: Node; // the parent block node
+  readonly block: Element; // the parent block node
```
Comment (spocke) [inline modules/tinymce/src/core/main/ts/textpatterns/core/PatternTypes.ts:56]: Maybe change the comment as well to `the parent block element`

File: modules/tinymce/src/core/test/ts/browser/textpatterns/TextPatternsPublicApiTest.ts [hunk 5]
```diff
@@ -0,0 +1,124 @@
+import { describe, it } from '@ephox/bedrock-client';
+import { Arr } from '@ephox/katamari';
+import { TinyAssertions, TinyHooks, TinyUiActions } from '@ephox/wrap-mcagar';
+
+// eslint-disable-next-line @tinymce/no-publicapi-module-imports
+import { Editor, PluginManager, TextPatterns } from 'tinymce/core/api/PublicApi';
+
+// The purpose of this test is to ensure that we have all the types required
+// for managing TextPatterns exposed through PublicApi
+describe('browser.tinymce.textpatterns.TextPatternsPublicApiTest', () => {
```
Comment (spocke) [inline modules/tinymce/src/core/test/ts/browser/textpatterns/TextPatternsPublicApiTest.ts:10]: I this the first time we have a test that only has the purpose of testing the types. Should these be a new category of tests or should type testing be else where?

**Review by spocke — APPROVED** [review]

File: modules/tinymce/src/core/test/ts/browser/textpatterns/TextPatternsPublicApiTest.ts (hunk 5 above)
Comment (lnewson) [reply modules/tinymce/src/core/test/ts/browser/textpatterns/TextPatternsPublicApiTest.ts:10]: It's actually been done before fwiw. Here's the first one I can think of: https://github.com/tinymce/tinymce/blob/b69c943324af5e127fe53f296f8ff1d4df0769b2/modules/katamari/src/test/ts/atomic/api/data/CovarianceAssertionTest.ts.

File: modules/tinymce/src/core/main/ts/api/textpatterns/TextPatterns.ts [hunk 6]
```diff
@@ -1,7 +1,15 @@
-import { RawPattern, RawDynamicPatternsLookup, DynamicPatternContext } from '../../textpatterns/core/PatternTypes';
+import { RawPattern, RawDynamicPatternsLookup, DynamicPatternContext, DynamicPatternsLookup, Pattern, BlockCmdPattern, BlockFormatPattern, InlineCmdPattern, InlinePattern, InlineFormatPattern, BlockPattern } from '../../textpatterns/core/PatternTypes';
```
[the remark this answers is not in the document]
Comment (lnewson) [reply modules/tinymce/src/core/main/ts/api/textpatterns/TextPatterns.ts:1]: We really should look at fixing all those max line length linter issues one day so we can re-enable the rule. Atm it's configured but it's really large (260 vs 160).

File: modules/tinymce/src/core/test/ts/browser/textpatterns/TextPatternsPublicApiTest.ts (hunk 5 above)
Comment (ephox-mogran) [reply modules/tinymce/src/core/test/ts/browser/textpatterns/TextPatternsPublicApiTest.ts:10]: I think testing the presence of types on the PublicApi in lots of places sounds like a good way to make sure we don't accidentally break something. Especially when some of the exports from PublicApi are "all the exports in this other module"

File: modules/tinymce/src/core/main/ts/api/textpatterns/TextPatterns.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (ephox-mogran) [reply modules/tinymce/src/core/main/ts/api/textpatterns/TextPatterns.ts:1]: What is the format for multilines. Is it 

```
import { ..., ...., ..., 
  ..., ..., } from '..'
```

or

```
import {
  ...., ...., ..., ....,
  ...., ...., ...., ....
} from '...'

or something else ?

File: modules/tinymce/src/core/main/ts/textpatterns/core/PatternTypes.ts (hunk 4 above)
Comment (ephox-mogran) [reply modules/tinymce/src/core/main/ts/textpatterns/core/PatternTypes.ts:56]: Good point. 

- [ ] change comment.

File: modules/tinymce/src/core/main/ts/api/textpatterns/TextPatterns.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (lnewson) [reply modules/tinymce/src/core/main/ts/api/textpatterns/TextPatterns.ts:1]: We really just need to either find a linter rule for this or write one... it's the latter anyways. So it's either a single line or if it's multiline then the import/from statements should be on their own lines.

[the remark this answers is not in the document]
Comment (TheSpyder) [reply modules/tinymce/src/core/main/ts/api/textpatterns/TextPatterns.ts:1]: yes I'm trying to put a stop to style questions in PRs. If there's a style issue, please just log a ticket to update (or add) a linter rule to cover it.

File: modules/tinymce/src/core/test/ts/browser/textpatterns/TextPatternsPublicApiTest.ts (hunk 5 above)
Comment (TheSpyder) [reply modules/tinymce/src/core/test/ts/browser/textpatterns/TextPatternsPublicApiTest.ts:10]: As long as refactoring doesn't auto-update the tests and make it pass incorrectly 😆

**Review by TheSpyder — APPROVED** [review]

File: modules/tinymce/src/core/test/ts/browser/textpatterns/TextPatternsPublicApiTest.ts (hunk 5 above)
Comment (ephox-mogran) [reply modules/tinymce/src/core/test/ts/browser/textpatterns/TextPatternsPublicApiTest.ts:10]: Well, then it's the review's job to see that the test changed :P

File: modules/tinymce/src/core/main/ts/api/textpatterns/TextPatterns.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (lnewson) [reply modules/tinymce/src/core/main/ts/api/textpatterns/TextPatterns.ts:1]: Found https://www.npmjs.com/package/eslint-plugin-import-newlines which I think should do what's needed here.

**Pull request merged**